### PR TITLE
Implement changes in io interfaces

### PR DESCRIPTION
**Pull request by zmichaels11**
Implemented changes in IO interfaces that were introduced in recent commits to rosbag2.

### Dependent PRs
- [x] ros2/rosbag2#184
- [x] ros2/rosbag2#183

### CODE REVIEW

File: rosbag2_bag_v2_plugins/src/rosbag2_bag_v2_plugins/storage/rosbag_v2_storage.hpp
Comment (dabonnie) [inline rosbag2_bag_v2_plugins/src/rosbag2_bag_v2_plugins/storage/rosbag_v2_storage.hpp:46]: Method doc?

File: rosbag2_bag_v2_plugins/src/rosbag2_bag_v2_plugins/storage/rosbag_v2_storage.hpp
Comment (dabonnie) [inline rosbag2_bag_v2_plugins/src/rosbag2_bag_v2_plugins/storage/rosbag_v2_storage.hpp:54]: Method doc?

File: rosbag2_bag_v2_plugins/src/rosbag2_bag_v2_plugins/storage/rosbag_v2_storage.cpp
Comment (dabonnie) [inline rosbag2_bag_v2_plugins/src/rosbag2_bag_v2_plugins/storage/rosbag_v2_storage.cpp]: Probably should be defined as a constant.

File: rosbag2_bag_v2_plugins/src/rosbag2_bag_v2_plugins/storage/rosbag_v2_storage.hpp
Comment (thomas-moulard) [inline rosbag2_bag_v2_plugins/src/rosbag2_bag_v2_plugins/storage/rosbag_v2_storage.hpp:46]:
```suggestion
  uint64_t get_bagfile_size() const final;
```

prefer `final` to `override` whenever possible.

File: rosbag2_bag_v2_plugins/src/rosbag2_bag_v2_plugins/storage/rosbag_v2_storage.hpp
Comment (thomas-moulard) [inline rosbag2_bag_v2_plugins/src/rosbag2_bag_v2_plugins/storage/rosbag_v2_storage.hpp:54]:
```suggestion
  std::string get_storage_identifier() const final;
```

File: rosbag2_bag_v2_plugins/test/rosbag2_bag_v2_plugins/test_rosbag_v2_storage.cpp
Comment (thomas-moulard) [inline rosbag2_bag_v2_plugins/test/rosbag2_bag_v2_plugins/test_rosbag_v2_storage.cpp:79]: (nit - feel fee to ignore)\

```suggestion
  EXPECT_EQ(bag_metadata.version, 2);
```

File: rosbag2_bag_v2_plugins/src/rosbag2_bag_v2_plugins/storage/rosbag_v2_storage.cpp
Comment (thomas-moulard) [reply rosbag2_bag_v2_plugins/src/rosbag2_bag_v2_plugins/storage/rosbag_v2_storage.cpp]: `static constexpr const char * const STORAGE_IDENTIFIER = "rosbag_v2";` at the top of this file.

File: rosbag2_bag_v2_plugins/src/rosbag2_bag_v2_plugins/storage/rosbag_v2_storage.hpp
Comment (zmichaels11) [reply rosbag2_bag_v2_plugins/src/rosbag2_bag_v2_plugins/storage/rosbag_v2_storage.hpp:54]: Added method doc for `get_storage_identifier`

File: rosbag2_bag_v2_plugins/test/rosbag2_bag_v2_plugins/test_rosbag_v2_storage.cpp
Comment (zmichaels11) [reply rosbag2_bag_v2_plugins/test/rosbag2_bag_v2_plugins/test_rosbag_v2_storage.cpp:79]: I'll open up an issue for replacing `EXPECT_THAT` with `EXPECT_EQ`

**Review by thomas-moulard — COMMENTED** [review]
While the code generally LGTM, how could we cope with `get_storage_identifier()` and `get_bagfile_size()` being unimplemented here before? Are we in the process of expending the API?

**Review by thomas-moulard — COMMENTED** [review]
Hmmm the C++ code guidelines do not agree with me.
https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#c139-use-final-sparingly
...so I guess override is ok.

File: rosbag2_bag_v2_plugins/src/rosbag2_bag_v2_plugins/storage/rosbag_v2_storage.hpp
Comment (thomas-moulard) [reply rosbag2_bag_v2_plugins/src/rosbag2_bag_v2_plugins/storage/rosbag_v2_storage.hpp:46]: this comment is ok to ignore see my PR comment

File: rosbag2_bag_v2_plugins/src/rosbag2_bag_v2_plugins/storage/rosbag_v2_storage.hpp
Comment (thomas-moulard) [reply rosbag2_bag_v2_plugins/src/rosbag2_bag_v2_plugins/storage/rosbag_v2_storage.hpp:54]: this comment is ok to ignore see my PR comment

**Comment by zmichaels11** [comment]
> While the code generally LGTM, how could we cope with `get_storage_identifier()` and `get_bagfile_size()` being unimplemented here before? Are we in the process of expending the API?

Yeah, both functions were added into the API in anticipation of supporting bagfile splitting. I'll link the merged PR numbers related to these changes.

**Comment by thomas-moulard** [comment]
:+1: Thanks!

**Review by thomas-moulard — APPROVED** [review]

File: rosbag2_bag_v2_plugins/src/rosbag2_bag_v2_plugins/storage/rosbag_v2_storage.cpp
Comment (dabonnie) [reply rosbag2_bag_v2_plugins/src/rosbag2_bag_v2_plugins/storage/rosbag_v2_storage.cpp]: It would be nice to get this in before the merge, so those who are calling get_storage_identifier don't have to check against a manually constructed string :-)

Comment (zmichaels11) [reply rosbag2_bag_v2_plugins/src/rosbag2_bag_v2_plugins/storage/rosbag_v2_storage.cpp]: I actually did:
```c++
namespace
{
const char * IDENTIFIER = "rosbag_v2";
}
```

I'll change it to  ```constexpr const char * const```

File: rosbag2_bag_v2_plugins/src/rosbag2_bag_v2_plugins/storage/rosbag_v2_storage.hpp
Comment (zmichaels11) [reply rosbag2_bag_v2_plugins/src/rosbag2_bag_v2_plugins/storage/rosbag_v2_storage.hpp:46]: This method behaves exactly like how the document on the interface behaves.
Is this fine as is? Otherwise I'd just be copying in the documentation from `rosbag2_storage`.

On the other method, I was able to make it more specific so its kind of a different case.

File: rosbag2_bag_v2_plugins/src/rosbag2_bag_v2_plugins/storage/rosbag_v2_storage.cpp
Comment (thomas-moulard) [inline rosbag2_bag_v2_plugins/src/rosbag2_bag_v2_plugins/storage/rosbag_v2_storage.cpp]:
```suggestion
constexpr const char * const IDENTIFIER = "rosbag_v2";
```

the ns is ok, as we discussed https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#sf22-use-an-unnamed-anonymous-namespace-for-all-internalnonexported-entities

**Comment by Karsten1987** [comment]
does it make sense to enhance this PR to comply with https://github.com/ros2/rosbag2/pull/184?

**Comment by zmichaels11** [comment]
> does it make sense to enhance this PR to comply with [ros2/rosbag2#184](https://github.com/ros2/rosbag2/pull/184)?

Yes, unless you want to merge this commit first.

**Comment by Karsten1987** [comment]
sounds good. Please go ahead and advance this PR with the changes needed and I we can run CI together on this.

**Comment by jacobperron** [comment]
FYI, I've tried to use the current state of this PR and get the following compilation error:

```
In file included from /home/jacob/ws/ros1_bridge_ws/install/class_loader/include/class_loader/class_loader_core.hpp:56:0,
                 from /home/jacob/ws/ros1_bridge_ws/install/class_loader/include/class_loader/class_loader.hpp:55,
                 from /home/jacob/ws/ros1_bridge_ws/install/class_loader/include/class_loader/multi_library_class_loader.hpp:52,
                 from /home/jacob/ws/ros1_bridge_ws/install/pluginlib/include/pluginlib/class_loader.hpp:58,
                 from /home/jacob/ws/rosbag2_bag_v2_ws/install/include/rosbag/bag.h:67,
                 from /home/jacob/ws/rosbag2_bag_v2_ws/src/rosbag2_bag_v2/rosbag2_bag_v2_plugins/src/rosbag2_bag_v2_plugins/storage/rosbag_v2_storage.hpp:23,
                 from /home/jacob/ws/rosbag2_bag_v2_ws/src/rosbag2_bag_v2/rosbag2_bag_v2_plugins/src/rosbag2_bag_v2_plugins/storage/rosbag_v2_storage.cpp:15:
/home/jacob/ws/ros1_bridge_ws/install/class_loader/include/class_loader/meta_object.hpp: In instantiation of ‘B* class_loader::impl::MetaObject<C, B>::create() const [with C = rosbag2_bag_v2_plugins::RosbagV2Storage; B = rosbag2_storage::storage_interfaces::ReadOnlyInterface]’:
/home/jacob/ws/rosbag2_bag_v2_ws/src/rosbag2_bag_v2/rosbag2_bag_v2_plugins/src/rosbag2_bag_v2_plugins/storage/rosbag_v2_storage.cpp:185:1:   required from here
/home/jacob/ws/ros1_bridge_ws/install/class_loader/include/class_loader/meta_object.hpp:193:12: error: invalid new-expression of abstract class type ‘rosbag2_bag_v2_plugins::RosbagV2Storage’
     return new C;
            ^~~~~
In file included from /home/jacob/ws/rosbag2_bag_v2_ws/src/rosbag2_bag_v2/rosbag2_bag_v2_plugins/src/rosbag2_bag_v2_plugins/storage/rosbag_v2_storage.cpp:15:0:
/home/jacob/ws/rosbag2_bag_v2_ws/src/rosbag2_bag_v2/rosbag2_bag_v2_plugins/src/rosbag2_bag_v2_plugins/storage/rosbag_v2_storage.hpp:29:7: note:   because the following virtual functions are pure within ‘rosbag2_bag_v2_plugins::RosbagV2Storage’:
 class RosbagV2Storage : public rosbag2_storage::storage_interfaces::ReadOnlyInterface
       ^~~~~~~~~~~~~~~
In file included from /home/jacob/ws/ros1_bridge_ws/install/rosbag2_storage/include/rosbag2_storage/storage_interfaces/read_only_interface.hpp:20:0,
                 from /home/jacob/ws/rosbag2_bag_v2_ws/src/rosbag2_bag_v2/rosbag2_bag_v2_plugins/src/rosbag2_bag_v2_plugins/storage/rosbag_v2_storage.hpp:22,
                 from /home/jacob/ws/rosbag2_bag_v2_ws/src/rosbag2_bag_v2/rosbag2_bag_v2_plugins/src/rosbag2_bag_v2_plugins/storage/rosbag_v2_storage.cpp:15:
/home/jacob/ws/ros1_bridge_ws/install/rosbag2_storage/include/rosbag2_storage/storage_interfaces/base_info_interface.hpp:41:23: note: 	virtual std::__cxx11::string rosbag2_storage::storage_interfaces::BaseInfoInterface::get_relative_path() const
   virtual std::string get_relative_path() const = 0;
                       ^~~~~~~~~~~~~~~~~
make[2]: *** [CMakeFiles/rosbag2_bag_v2_plugins.dir/src/rosbag2_bag_v2_plugins/storage/rosbag_v2_storage.cpp.o] Error 1
make[2]: *** Waiting for unfinished jobs....
make[1]: *** [CMakeFiles/rosbag2_bag_v2_plugins.dir/all] Error 2
make[1]: *** Waiting for unfinished jobs....
make: *** [all] Error 2
```

**Comment by zmichaels11** [comment]
@jacobperron just resolved that issue; it was due to recent API changes introduced in ros2/rosbag2#184.

@Karsten1987 This PR is ready

**Comment by jacobperron** [comment]
@zmichaels11 Yep, thanks. I'm able to compile now.

**Comment by zmichaels11** [comment]
@Karsten1987 is there anything blocking this PR from merging?

**Review by Karsten1987 — APPROVED** [review]

**Pull request merged**
